File: bindings/csharp/tests/parity/Runner.cs
// Copyright (c) 2026 Capgemini Engineering Research and Development.
//
// This file is part of OCCT-Light software library.
//
// This library is free software; you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License version 3 as published
// by the Free Software Foundation, with an option to use any later version.
// Consult the file LICENSE_AGPL_30.txt included in OCCT-Light distribution
// for complete text of the license and disclaimer of any warranty.
//
// Alternatively, this file may be used under the terms of a commercial
// license or contractual agreement.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using OcctL;
using OcctL.Native;
using OcctL.Options;

namespace OcctL.Parity;

/// <summary>
/// Parity runner. Reads one scenario from <c>tests/binding_parity</c>, executes it
/// through the idiomatic C# facade, checks the expected fields, and prints one
/// canonical JSON object.
/// </summary>
public static class Runner
{
    public static int Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.Error.WriteLine("usage: OcctL.Parity <scenario.json>");
            return 2;
        }

        string aPath = args[0];
        if (!File.Exists(aPath))
        {
            Console.Error.WriteLine($"Scenario not found: {aPath}");
            return 2;
        }

        using var aStream = File.OpenRead(aPath);
        using var aDoc = JsonDocument.Parse(aStream);
        JsonElement aRoot = aDoc.RootElement;
        string aScenario = aRoot.GetProperty("scenario").GetString() ?? string.Empty;
        IReadOnlyDictionary<string, object> anActual = RunScenario(aScenario, aRoot);
        bool aMatches = Matches(aRoot.GetProperty("expected"), anActual);

        string aJson = JsonSerializer.Serialize(new
        {
            scenario = aScenario,
            binding = "csharp",
            actual = anActual,
            matches = aMatches,
        });
        Console.WriteLine(aJson);
        return aMatches ? 0 : 1;
    }

    private static IReadOnlyDictionary<string, object> RunScenario(string scenario, JsonElement doc)
        => scenario switch
        {
            "build_box" => RunBuildBox(doc),
            "fuse_two_boxes" => RunFuseTwoBoxes(),
            "cut_box_corner" => RunCutBoxCorner(),
            "common_two_overlapping_boxes" => RunCommonTwoOverlappingBoxes(),
            "section_two_overlapping_boxes" => RunSectionTwoOverlappingBoxes(),
            "split_box_by_box" => RunSplitBoxByBox(),
            "history_modified_after_fuse" => RunHistoryModifiedAfterFuse(),
            _ => throw new InvalidOperationException($"Unsupported parity scenario: {scenario}"),
        };

    private static IReadOnlyDictionary<string, object> RunBuildBox(JsonElement doc)
    {
        double aDx = 10.0;
        double aDy = 10.0;
        double aDz = 5.0;
        if (doc.TryGetProperty("params", out JsonElement aParams))
        {
            if (aParams.TryGetProperty("dx", out JsonElement aDxValue)) aDx = aDxValue.GetDouble();
            if (aParams.TryGetProperty("dy", out JsonElement aDyValue)) aDy = aDyValue.GetDouble();
            if (aParams.TryGetProperty("dz", out JsonElement aDzValue)) aDz = aDzValue.GetDouble();
        }

        using var aGraph = new Graph();
        Prim.MakeBox(aGraph, aDx, aDy, aDz);
        return new Dictionary<string, object>
        {
            ["face_count"] = aGraph.FaceCount,
            ["edge_count"] = aGraph.EdgeCount,
            ["vertex_count"] = aGraph.VertexCount,
            ["solid_count"] = aGraph.SolidCount,
        };
    }

    private static IReadOnlyDictionary<string, object> RunFuseTwoBoxes()
    {
        using var aGraph = new Graph();
        NodeId aBoxA = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aBoxB = Prim.MakeBox(aGraph, Box(10.0, 10.0, 10.0, 5.0, 0.0, 0.0));
        Uid[] aBoxAFaceUids = aGraph.Faces().Take(6).Select(aGraph.UidOf).ToArray();

        NodeId aResult = Bool.Fuse(aGraph, [aBoxA], [aBoxB]);
        bool aHistoryNonEmpty = aBoxAFaceUids.Any(uid => aGraph.HistoryModified(uid).Length != 0
                                                      || aGraph.HistoryGenerated(uid).Length != 0);

        return new Dictionary<string, object>
        {
            ["root_kind"] = KindString(aGraph.NodeKind(aResult)),
            ["history_modified_nonempty_on_first_face_of_box_a"] = aHistoryNonEmpty,
        };
    }

    private static IReadOnlyDictionary<string, object> RunCutBoxCorner()
    {
        using var aGraph = new Graph();
        NodeId aBox = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aTool = Prim.MakeBox(aGraph, Box(5.0, 5.0, 5.0, 7.5, 7.5, 7.5));
        NodeId aResult = Bool.Cut(aGraph, [aBox], [aTool]);

        return new Dictionary<string, object>
        {
            ["root_kind"] = KindString(aGraph.NodeKind(aResult)),
            ["solid_count"] = aGraph.SolidCount,
        };
    }

    private static IReadOnlyDictionary<string, object> RunCommonTwoOverlappingBoxes()
    {
        using var aGraph = new Graph();
        NodeId aBoxA = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aBoxB = Prim.MakeBox(aGraph, Box(10.0, 10.0, 10.0, 5.0, 5.0, 5.0));
        NodeId aResult = Bool.Common(aGraph, [aBoxA], [aBoxB]);

        return new Dictionary<string, object>
        {
            ["root_kind"] = KindString(aGraph.NodeKind(aResult)),
            ["solid_count"] = aGraph.SolidCount,
        };
    }

    private static IReadOnlyDictionary<string, object> RunSectionTwoOverlappingBoxes()
    {
        using var aGraph = new Graph();
        NodeId aBoxA = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aBoxB = Prim.MakeBox(aGraph, Box(10.0, 10.0, 10.0, 5.0, 0.0, 0.0));
        int aEdgesBefore = aGraph.EdgeCount;
        NodeId aResult = Bool.Section(aGraph, [aBoxA], [aBoxB]);

        return new Dictionary<string, object>
        {
            ["root_kind"] = KindString(aGraph.NodeKind(aResult)),
            ["edge_count_increased"] = aGraph.EdgeCount > aEdgesBefore,
        };
    }

    private static IReadOnlyDictionary<string, object> RunSplitBoxByBox()
    {
        using var aGraph = new Graph();
        NodeId aBoxA = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aBoxB = Prim.MakeBox(aGraph, Box(10.0, 10.0, 10.0, 0.0, 0.0, 5.0));
        NodeId aResult = Bool.Split(aGraph, [aBoxA], [aBoxB]);

        return new Dictionary<string, object>
        {
            ["root_kind"] = KindString(aGraph.NodeKind(aResult)),
            ["compound_count"] = aGraph.CompoundCount,
            ["solid_count"] = aGraph.SolidCount,
        };
    }

    private static IReadOnlyDictionary<string, object> RunHistoryModifiedAfterFuse()
    {
        using var aGraph = new Graph();
        NodeId aBoxA = Prim.MakeBox(aGraph, 10.0, 10.0, 10.0);
        NodeId aBoxB = Prim.MakeBox(aGraph, Box(10.0, 10.0, 10.0, 5.0, 0.0, 0.0));
        Uid[] aFaceUids = aGraph.Faces().Take(12).Select(aGraph.UidOf).ToArray();
        Uid[] aBoxAFaceUids = aFaceUids[0..6];
        Uid[] aBoxBFaceUids = aFaceUids[6..12];

        NodeId aResult = Bool.Fuse(aGraph, [aBoxA], [aBoxB]);
        bool aBoxAModified = AnyModified(aGraph, aBoxAFaceUids);
        bool aBoxBModified = AnyModified(aGraph, aBoxBFaceUids);

        return new Dictionary<string, object>
        {
            ["box_a_modified_nonempty"] = aBoxAModified,
            ["box_b_modified_nonempty"] = aBoxBModified,
        };
    }

    private static bool AnyModified(Graph graph, IEnumerable<Uid> uids)
        => uids.Any(uid => graph.HistoryModified(uid).Length != 0 || graph.HistoryGenerated(uid).Length != 0);

    private static PrimBoxInfo Box(double dx, double dy, double dz, double x, double y, double z)
        => new()
        {
            Dx = dx,
            Dy = dy,
            Dz = dz,
            Placement = Placement(x, y, z),
        };

    private static OcctlAxis2Placement Placement(double x, double y, double z)
        => new()
        {
            Location = new OcctlPoint3 { X = x, Y = y, Z = z },
            XDir = new OcctlDirection3 { X = 1.0, Y = 0.0, Z = 0.0 },
            XDirRef = new OcctlDirection3 { X = 0.0, Y = 1.0, Z = 0.0 },
        };

    private static string KindString(OcctlNodeKind kind)
        => kind switch
        {
            OcctlNodeKind.KindSolid => "solid",
            OcctlNodeKind.KindShell => "shell",
            OcctlNodeKind.KindFace => "face",
            OcctlNodeKind.KindWire => "wire",
            OcctlNodeKind.KindEdge => "edge",
            OcctlNodeKind.KindVertex => "vertex",
            OcctlNodeKind.KindCompound => "compound",
            OcctlNodeKind.KindCompsolid => "compsolid",
            _ => "unknown",
        };

    private static bool Matches(JsonElement expected, IReadOnlyDictionary<string, object> actual)
    {
        foreach (JsonProperty aProperty in expected.EnumerateObject())
        {
            string aName = aProperty.Name;
            if (aName.EndsWith("_at_least", StringComparison.Ordinal))
            {
                string anActualName = aName[..^"_at_least".Length];
                if (!actual.TryGetValue(anActualName, out object? anActualValue)
                    || Convert.ToInt32(anActualValue) < aProperty.Value.GetInt32())
                {
                    return false;
                }
                continue;
            }

            if (aName.EndsWith("_in", StringComparison.Ordinal))
            {
                string anActualName = aName[..^"_in".Length];
                if (!actual.TryGetValue(anActualName, out object? anActualValue)
                    || !aProperty.Value.EnumerateArray().Any(v => v.GetString() == Convert.ToString(anActualValue)))
                {
                    return false;
                }
                continue;
            }

            if (!actual.TryGetValue(aName, out object? aValue) || !MatchesScalar(aProperty.Value, aValue))
            {
                return false;
            }
        }
        return true;
    }

    private static bool MatchesScalar(JsonElement expected, object actual)
        => expected.ValueKind switch
        {
            JsonValueKind.True => actual is bool b && b,
            JsonValueKind.False => actual is bool b && !b,
            JsonValueKind.Number => Convert.ToDouble(actual) == expected.GetDouble(),
            JsonValueKind.String => Convert.ToString(actual) == expected.GetString(),
            _ => false,
        };
}

File: bindings/python/tests/parity/runner.py
#!/usr/bin/env python3
# Copyright (c) 2026 Capgemini Engineering Research and Development.
#
# This file is part of OCCT-Light software library.
#
# This library is free software; you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License version 3 as published
# by the Free Software Foundation, with an option to use any later version.
# Consult the file LICENSE_AGPL_30.txt included in OCCT-Light distribution
# for complete text of the license and disclaimer of any warranty.
#
# Alternatively, this file may be used under the terms of a commercial
# license or contractual agreement.
#
# SPDX-License-Identifier: AGPL-3.0-or-later

"""Parity runner: execute a binding_parity scenario and print canonical JSON.

Usage::

    python tests/parity/runner.py tests/binding_parity/build_box.json

The script reads a scenario file from ``tests/binding_parity/``, executes the
equivalent operations through :mod:`occtl`, and writes one JSON object to
stdout with the same schema every binding's runner emits. Phase 2 compares
the four bindings' outputs byte-for-byte.

Scenario shape::

    {
      "scenario": "build_box",
      "description": "...",
      "expected": {
        "face_count": 6,
        "edge_count": 12,
        "vertex_count": 8,
        "solid_count": 1
      },
      // optional:
      "params": { "dx": 10.0, "dy": 10.0, "dz": 5.0 }
    }

Output schema::

    {
      "scenario": "build_box",
      "binding": "python",
      "actual": { "face_count": 6, "edge_count": 12,
                  "vertex_count": 8, "solid_count": 1 },
      "matches": true
    }
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path


def _run_build_box(scenario: dict) -> dict:
    import occtl

    params = scenario.get("params") or {}
    dx = float(params.get("dx", 10.0))
    dy = float(params.get("dy", 10.0))
    dz = float(params.get("dz", 5.0))

    with occtl.Graph() as g:
        occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=dx, dy=dy, dz=dz))
        return {
            "face_count": g.face_count,
            "edge_count": g.edge_count,
            "vertex_count": g.vertex_count,
            "solid_count": g.solid_count,
        }


def _run_fuse_two_boxes(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        box_a = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        shifted = occtl.Axis2Placement(
            location=occtl.Point3(5.0, 0.0, 0.0),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        box_b = occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0, placement=shifted),
        )

        # Capture box A's face UIDs before fuse so they survive the result
        # being mixed in graph-wide iteration.
        face_uids_before = [g.uid_of(f) for f in g.face_iter()]
        a_face_uids = face_uids_before[0:6]

        result = occtl.bool_.fuse(g, [box_a], [box_b])

        history_modified_nonempty = False
        for uid in a_face_uids:
            if g.history_modified(uid) or g.history_generated(uid):
                history_modified_nonempty = True
                break

        return {
            "root_kind": _kind_str(g, result),
            "history_modified_nonempty_on_first_face_of_box_a": history_modified_nonempty,
        }


def _kind_str(graph, node_id) -> str:
    """Map an occtl_node_kind_t into the short string used in scenarios."""
    from occtl._generated._raw import lib, ffi
    out = ffi.new("occtl_node_kind_t*")
    raw = ffi.new("occtl_node_id_t*", [node_id.bits])[0]
    lib.occtl_graph_node_kind(graph._as_ptr(), raw, out)
    int_to_str = {
        int(lib.OCCTL_KIND_SOLID):     "solid",
        int(lib.OCCTL_KIND_SHELL):     "shell",
        int(lib.OCCTL_KIND_FACE):      "face",
        int(lib.OCCTL_KIND_WIRE):      "wire",
        int(lib.OCCTL_KIND_EDGE):      "edge",
        int(lib.OCCTL_KIND_VERTEX):    "vertex",
        int(lib.OCCTL_KIND_COMPOUND):  "compound",
        int(lib.OCCTL_KIND_COMPSOLID): "compsolid",
    }
    return int_to_str.get(int(out[0]), "unknown")


def _run_cut_box_corner(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        box = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        corner = occtl.Axis2Placement(
            location=occtl.Point3(7.5, 7.5, 7.5),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=5.0, dy=5.0, dz=5.0, placement=corner),
        )
        result = occtl.bool_.cut(g, [box], list(g.solid_iter())[1:])  # tool is the second solid
        return {
            "root_kind": _kind_str(g, result),
            "solid_count": g.solid_count,
        }


def _run_common_two_overlapping_boxes(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        a = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        shifted = occtl.Axis2Placement(
            location=occtl.Point3(5.0, 5.0, 5.0),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        b = occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0, placement=shifted),
        )
        result = occtl.bool_.common(g, [a], [b])
        return {
            "root_kind": _kind_str(g, result),
            "solid_count": g.solid_count,
        }


def _run_section_two_overlapping_boxes(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        a = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        shifted = occtl.Axis2Placement(
            location=occtl.Point3(5.0, 0.0, 0.0),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        b = occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0, placement=shifted),
        )
        edge_count_before = g.edge_count
        result = occtl.bool_.section(g, [a], [b])
        return {
            "root_kind": _kind_str(g, result),
            "edge_count_increased": g.edge_count > edge_count_before,
        }


def _run_split_box_by_box(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        a = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        shifted = occtl.Axis2Placement(
            location=occtl.Point3(0.0, 0.0, 5.0),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        b = occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0, placement=shifted),
        )
        result = occtl.bool_.split(g, [a], [b])
        return {
            "root_kind": _kind_str(g, result),
            "compound_count": g.compound_count,
            "solid_count": g.solid_count,
        }


def _run_history_modified_after_fuse(scenario: dict) -> dict:
    import occtl

    with occtl.Graph() as g:
        a = occtl.prim.make_box(g, occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0))
        shifted = occtl.Axis2Placement(
            location=occtl.Point3(5.0, 0.0, 0.0),
            x_dir=occtl.Direction3(1.0, 0.0, 0.0),
            x_dir_ref=occtl.Direction3(0.0, 1.0, 0.0),
        )
        b = occtl.prim.make_box(
            g,
            occtl.prim.BoxInfo(dx=10.0, dy=10.0, dz=10.0, placement=shifted),
        )
        # Capture the face UIDs of each input box BEFORE the fuse. The
        # fuse adds new faces (the result solid's faces) so iterating
        # face_iter() afterwards mixes inputs and results.
        faces_before = [g.uid_of(f) for f in g.face_iter()]
        # Tag-by-position: first 6 belong to a, next 6 to b (box-shape
        # primitives emit 6 faces in deterministic order).
        a_face_uids = faces_before[0:6]
        b_face_uids = faces_before[6:12]

        occtl.bool_.fuse(g, [a], [b])

        def any_modified(uids):
            for uid in uids:
                if g.history_modified(uid) or g.history_generated(uid):
                    return True
            return False

        return {
            "box_a_modified_nonempty": any_modified(a_face_uids),
            "box_b_modified_nonempty": any_modified(b_face_uids),
        }


_HANDLERS = {
    "build_box": _run_build_box,
    "fuse_two_boxes": _run_fuse_two_boxes,
    "cut_box_corner": _run_cut_box_corner,
    "common_two_overlapping_boxes": _run_common_two_overlapping_boxes,
    "section_two_overlapping_boxes": _run_section_two_overlapping_boxes,
    "split_box_by_box": _run_split_box_by_box,
    "history_modified_after_fuse": _run_history_modified_after_fuse,
}


def main(argv: "list[str] | None" = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("scenario", type=Path)
    args = parser.parse_args(argv)

    scenario = json.loads(args.scenario.read_text(encoding="utf-8"))
    handler = _HANDLERS.get(scenario["scenario"])
    if handler is None:
        sys.stderr.write(
            f"runner: unknown scenario '{scenario['scenario']}'. "
            f"Known: {sorted(_HANDLERS)}\n"
        )
        return 2

    actual = handler(scenario)
    expected = scenario.get("expected") or {}
    output = {
        "scenario": scenario["scenario"],
        "binding": "python",
        "actual": actual,
        "matches": _matches(actual, expected),
    }
    sys.stdout.write(json.dumps(output, sort_keys=True) + "\n")
    return 0 if output["matches"] else 1


def _matches(actual: dict, expected: dict) -> bool:
    """Match an actual result against an expected dict.

    Each expected key drives a comparator:
      * key ``X_in`` ⇒ actual[X] must equal one of expected[X_in].
      * key ``X_at_least`` ⇒ actual[X] >= expected[X_at_least].
      * key ``X_greater_than`` ⇒ actual[X] > expected[X_greater_than].
      * otherwise ⇒ actual[X] must equal expected[X] exactly.
    """
    for key, want in expected.items():
        if key.endswith("_in"):
            base = key[:-len("_in")]
            if actual.get(base) not in want:
                return False
        elif key.endswith("_at_least"):
            base = key[:-len("_at_least")]
            if not (actual.get(base, 0) >= want):
                return False
        elif key.endswith("_greater_than"):
            base = key[:-len("_greater_than")]
            if not (actual.get(base, 0) > want):
                return False
        else:
            if actual.get(key) != want:
                return False
    return True


if __name__ == "__main__":
    raise SystemExit(main())
